File: labs/control-ui/api-shadow/entropy_scrambled_tunnel.ts
export class EntropyScrambledTunnel {
  public minPadding: number;
  public maxPadding: number;
  private secretKey: Uint8Array;
  constructor(secretKey: Uint8Array, minPad: number = 8, maxPad: number = 32) {
    this.secretKey = secretKey;
    this.minPadding = Math.max(4, minPad);
    this.maxPadding = Math.max(this.minPadding + 8, maxPad);
  }

  scramblePacket(payload: Uint8Array, seed: number): Uint8Array {
    const padRange = this.maxPadding - this.minPadding;
    const padLen = this.minPadding + (Math.abs(seed) % padRange);
    const totalLen = 4 + payload.length + padLen;

    const frame = new Uint8Array(totalLen);
    const view = new DataView(frame.buffer, frame.byteOffset, frame.byteLength);
    view.setUint16(0, payload.length);
    view.setUint16(2, padLen);
    frame.set(payload, 4);

    for (let i = 0; i < padLen; i++) {
      frame[4 + payload.length + i] = Math.floor(Math.random() * 256);
    }

    this.applyMask(frame, seed);
    return frame;
  }

  descramblePacket(scrambled: Uint8Array, seed: number): Uint8Array | null {
    if (scrambled.length < 4) return null;
    const unmasked = new Uint8Array(scrambled);
    this.applyMask(unmasked, seed);

    const view = new DataView(unmasked.buffer, unmasked.byteOffset, unmasked.byteLength);
    const payloadLen = view.getUint16(0);
    const padLen = view.getUint16(2);

    if (unmasked.length < 4 + payloadLen + padLen) return null;
    return unmasked.slice(4, 4 + payloadLen);
  }

  private applyMask(data: Uint8Array, seed: number): void {
    let s = BigInt(seed);
    const multiplier = 6364136223846793005n;
    const increment = 1442695040888963407n;
    for (let i = 0; i < data.length; i++) {
      s = (s * multiplier + increment) & 0xffffffffffffffffn;
      const mask = Number((s >> 33n) & 0xffn) ^ this.secretKey[i % this.secretKey.length]!;
      data[i]! ^= mask;
    }
  }

  static calculateEntropy(data: Uint8Array): number {
    if (data.length === 0) return 0;
    const counts = new Array(256).fill(0);
    for (let i = 0; i < data.length; i++) {
      counts[data[i]!]++;
    }
    const total = data.length;
    let entropy = 0;
    for (let i = 0; i < 256; i++) {
      if (counts[i] > 0) {
        const p = counts[i] / total;
        entropy -= p * Math.log2(p);
      }
    }
    return entropy;
  }
}

File: labs/mobile/donor-archive/tunnel/EntropyScrambledTunnel.kt
package com.luminet.android.tunnel

import java.nio.ByteBuffer
import java.security.SecureRandom
import kotlin.math.ln
import kotlin.math.max

class EntropyScrambledTunnel(
    private val secretKey: ByteArray,
    minPad: Int = 8,
    maxPad: Int = 32
) {
    private val minPadding = max(4, minPad)
    private val maxPadding = max(minPadding + 8, maxPad)
    private val random = SecureRandom()

    fun scramblePacket(payload: ByteArray, seed: Long): ByteArray {
        val padRange = maxPadding - minPadding
        val padLen = minPadding + (random.nextInt(padRange))

        val totalLen = 4 + payload.size + padLen
        val buf = ByteBuffer.allocate(totalLen)
        buf.putShort(payload.size.toShort())
        buf.putShort(padLen.toShort())
        buf.put(payload)

        val pad = ByteArray(padLen)
        random.nextBytes(pad)
        buf.put(pad)

        val raw = buf.array()
        applyMask(raw, seed)
        return raw
    }

    fun descramblePacket(scrambled: ByteArray, seed: Long): ByteArray? {
        if (scrambled.size < 4) return null
        val unmasked = scrambled.clone()
        applyMask(unmasked, seed)

        val buf = ByteBuffer.wrap(unmasked)
        val pLen = buf.getShort().toInt() and 0xFFFF
        val padLen = buf.getShort().toInt() and 0xFFFF

        if (unmasked.size < 4 + pLen + padLen) return null
        val payload = ByteArray(pLen)
        buf.get(payload)
        return payload
    }

    private fun applyMask(data: ByteArray, seed: Long) {
        var s = seed
        for (i in data.indices) {
            // Pseudo-random LCG stream from seed and secret key
            s = (s * 6364136223846793005L + 1442695040888963407L)
            val streamByte = ((s shr 33) xor secretKey[i % secretKey.size].toLong()).toByte()
            data[i] = (data[i].toInt() xor streamByte.toInt()).toByte()
        }
    }

    companion object {
        fun calculateShannonEntropy(data: ByteArray): Double {
            if (data.isEmpty()) return 0.0
            val counts = IntArray(256)
            for (b in data) {
                counts[b.toInt() and 0xFF]++
            }
            val total = data.size.toDouble()
            var entropy = 0.0
            val log2 = ln(2.0)
            for (c in counts) {
                if (c > 0) {
                    val p = c / total
                    entropy -= p * (ln(p) / log2)
                }
            }
            return entropy
        }
    }
}
